[JSC] WASM IPInt SIMD: add support for v128 arguments and results in IPInt call and tail-call instructions

dhecht opened the pull request:
#### 73f0c9d430cbc013ee54681b495425edd921da0c
<pre>
[JSC] WASM IPInt SIMD: add support for v128 arguments and results in IPInt call and tail-call instructions
<a href="https://bugs.webkit.org/show_bug.cgi?id=299606">https://bugs.webkit.org/show_bug.cgi?id=299606</a>
<a href="https://rdar.apple.com/161412501">rdar://161412501</a>

Reviewed by Yusuke Suzuki.

Supporting v128 call/tail-call argument and results requires removing
the various assumptions that IPInt makes that callframe arguments are
fixed sizes, as follows:

* CallInformation abstraction changes:

- CallInformation was exposing the number of argument/result stack
values in the callframe. WasmIPIntGenerator was using this information
and assuming fixed size arguments/results. This breaks the abstraction
since native callframes have variable sized elements. Instead, expose
the byte sizes of the header region and total size (already exposed).
These are redundant with the argument and result location lists but
are a useful summary for the clients.

* Bytecode generation updates:

- Update the bytecode generators to not assume uniform argument/result
sizes. Add various asserts that verifies that the bytecode metadata
and interpreters are in agreement and will tell us what &quot;small&quot;
bytecode needs to be updated if/when we change the wasm calling
convention. We could make the bytecode more generic but I didn&apos;t do
that in this change to minimize risk.

- m_highestReturnStackOffset was storing address of the top return
value. Now, m_topOfReturnStackFPOffset stores the address of the first
byte above the top return
value.

- The trailing metadata for mINT tail call now stores the size of the
argument space rather than the number of arguments, and the
corresponding loop in ipint_perform_tail_call that copies this region
is updated.

* Bytecode interpreter updates:

- All the &quot;small&quot; interpreter bytecodes that operate on FP registers
now operate on the full 128-bits rather than 64-bits.

- Introduce argumINT, uINT, and mINT bytecodes to handle vectors.
Also rename existing ones for clarity and consistency. Also make
bytecode enum values and handlers similarly named for clarity.

- Fix a hardcoded uintDispatch bounds check.

* Stack management fixes:

- The uINT destination stack (for copying results into the caller&apos;s
callframe when returning) was an empty descending stack, meaning each
bytecode allocated the space for the next result, assuming each result
is 8-bytes. But since callframe results aren&apos;t fixed size, change this
to a full descending stack so that each bytecode allocates its own
stack space since each bytecode knows the size of its result.

* Misc cleanup:

- Don&apos;t skip the WASM stack size asserts when useWasmIPIntSIMD is enabled.

- Factor redundant mINT argument bytecode generators into a single generator.

* Testing:

- Add a table driven test to make it easier to verify call and
tail-call frame construction cases. Each test case specifies the
signature of the callee and how callee results should be constructed
based on the arguments, and then the test harness generates the WAT
for the test case. I plan to extend this later to add more dimensions
to the test matrix, for example, to explicitly cover calling between
tiers and the indirect/ref call types.

Test: JSTests/wasm/stress/simd-instructions-calls.js
* JSTests/wasm/stress/simd-instructions-calls.js: Added.
(logV):
(i.i.2.0.i.10.5.5):
(resultMapping.new.Array.fill.map):
(generateSignature):
* Source/JavaScriptCore/llint/InPlaceInterpreter.h:
* Source/JavaScriptCore/llint/InPlaceInterpreter32_64.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::IPIntCallee::IPIntCallee):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmCallingConvention.h:
(JSC::Wasm::CallInformation::CallInformation):
(JSC::Wasm::WasmCallingConvention::callInformationFor const):
(JSC::Wasm::JSCallingConvention::callInformationFor const):
(JSC::Wasm::WasmCallingConvention::numberOfStackArguments const): Deleted.
(JSC::Wasm::WasmCallingConvention::numberOfStackValues const): Deleted.
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::addReturnData):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::assertAboutStackSize):
(JSC::Wasm::IPIntGenerator::addArguments):
(JSC::Wasm::addCallArgumentBytecode):
(JSC::Wasm::addCallResultBytecode):
(JSC::Wasm::IPIntGenerator::addCallCommonData):
(JSC::Wasm::IPIntGenerator::addTailCallCommonData):
(JSC::Wasm::IPIntGenerator::addCall):
(JSC::Wasm::IPIntGenerator::addCallIndirect):
(JSC::Wasm::IPIntGenerator::addCallRef):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.h:

Canonical link: <a href="https://commits.webkit.org/300695@main">https://commits.webkit.org/300695@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65c01a1a73f311c672990ac0a697def4ff816fee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123276 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42990 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33687 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129984 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75389 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d01bd61d-bdfd-42a9-9cad-b6d72a54afc0) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43714 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51585 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93706 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62163 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cda834c1-9866-41c7-a06c-0a16bb287571) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126229 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34831 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110301 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74334 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3fa109c4-a5bb-4aeb-a104-02d29598c070) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33796 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28460 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73502 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/115435 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104545 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28686 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132702 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/121807 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50226 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38218 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102196 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50602 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106525 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102051 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25990 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47402 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25624 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47000 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50081 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55842 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/152162 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49552 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38883 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52902 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51230 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/93278d6a470779d75f5d70cabf985d94daf46b36. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51388

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/65c01a1a73f311c672990ac0a697def4ff816fee. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51388

dhecht committed:
[JSC] WASM IPInt SIMD: add support for v128 arguments and results in IPInt call and tail-call instructions
https://bugs.webkit.org/show_bug.cgi?id=299606
rdar://161412501

Reviewed by Yusuke Suzuki.

Supporting v128 call/tail-call argument and results requires removing
the various assumptions that IPInt makes that callframe arguments are
fixed sizes, as follows:

* CallInformation abstraction changes:

- CallInformation was exposing the number of argument/result stack
values in the callframe. WasmIPIntGenerator was using this information
and assuming fixed size arguments/results. This breaks the abstraction
since native callframes have variable sized elements. Instead, expose
the byte sizes of the header region and total size (already exposed).
These are redundant with the argument and result location lists but
are a useful summary for the clients.

* Bytecode generation updates:

- Update the bytecode generators to not assume uniform argument/result
sizes. Add various asserts that verifies that the bytecode metadata
and interpreters are in agreement and will tell us what "small"
bytecode needs to be updated if/when we change the wasm calling
convention. We could make the bytecode more generic but I didn't do
that in this change to minimize risk.

- m_highestReturnStackOffset was storing address of the top return
value. Now, m_topOfReturnStackFPOffset stores the address of the first
byte above the top return
value.

- The trailing metadata for mINT tail call now stores the size of the
argument space rather than the number of arguments, and the
corresponding loop in ipint_perform_tail_call that copies this region
is updated.

* Bytecode interpreter updates:

- All the "small" interpreter bytecodes that operate on FP registers
now operate on the full 128-bits rather than 64-bits.

- Introduce argumINT, uINT, and mINT bytecodes to handle vectors.
Also rename existing ones for clarity and consistency. Also make
bytecode enum values and handlers similarly named for clarity.

- Fix a hardcoded uintDispatch bounds check.

* Stack management fixes:

- The uINT destination stack (for copying results into the caller's
callframe when returning) was an empty descending stack, meaning each
bytecode allocated the space for the next result, assuming each result
is 8-bytes. But since callframe results aren't fixed size, change this
to a full descending stack so that each bytecode allocates its own
stack space since each bytecode knows the size of its result.

* Misc cleanup:

- Don't skip the WASM stack size asserts when useWasmIPIntSIMD is enabled.

- Factor redundant mINT argument bytecode generators into a single generator.

* Testing:

- Add a table driven test to make it easier to verify call and
tail-call frame construction cases. Each test case specifies the
signature of the callee and how callee results should be constructed
based on the arguments, and then the test harness generates the WAT
for the test case. I plan to extend this later to add more dimensions
to the test matrix, for example, to explicitly cover calling between
tiers and the indirect/ref call types.

Test: JSTests/wasm/stress/simd-instructions-calls.js
* JSTests/wasm/stress/simd-instructions-calls.js: Added.
(logV):
(i.i.2.0.i.10.5.5):
(resultMapping.new.Array.fill.map):
(generateSignature):
* Source/JavaScriptCore/llint/InPlaceInterpreter.h:
* Source/JavaScriptCore/llint/InPlaceInterpreter32_64.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::IPIntCallee::IPIntCallee):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmCallingConvention.h:
(JSC::Wasm::CallInformation::CallInformation):
(JSC::Wasm::WasmCallingConvention::callInformationFor const):
(JSC::Wasm::JSCallingConvention::callInformationFor const):
(JSC::Wasm::WasmCallingConvention::numberOfStackArguments const): Deleted.
(JSC::Wasm::WasmCallingConvention::numberOfStackValues const): Deleted.
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::addReturnData):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::assertAboutStackSize):
(JSC::Wasm::IPIntGenerator::addArguments):
(JSC::Wasm::addCallArgumentBytecode):
(JSC::Wasm::addCallResultBytecode):
(JSC::Wasm::IPIntGenerator::addCallCommonData):
(JSC::Wasm::IPIntGenerator::addTailCallCommonData):
(JSC::Wasm::IPIntGenerator::addCall):
(JSC::Wasm::IPIntGenerator::addCallIndirect):
(JSC::Wasm::IPIntGenerator::addCallRef):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.h:

Canonical link: https://commits.webkit.org/300695@main
